src/adapter/src/client.rs

```diff
@@ -599,15 +601,26 @@ impl SessionClient {
         rows: Vec<Row>,
         ctx_extra: ExecuteContextExtra,
     ) -> Result<ExecuteResponse, AdapterError> {
-        self.send(|tx, session| Command::CopyRows {
-            id,
-            columns,
-            rows,
-            session,
-            tx,
-            ctx_extra,
-        })
-        .await
+        let catalog = self.catalog_snapshot().await;
+        // TODO: Remove this clone once we always have the session. It's currently needed because
+        // self.session returns a mut ref, so we can't call it twice.
+        let pcx = self.session().pcx().clone();
+        let conn_catalog = catalog.for_session(self.session());
+        let result: Result<_, AdapterError> =
+            mz_sql::plan::plan_copy_from(&pcx, &conn_catalog, id, columns, rows)
+                .err_into()
+                .and_then(|values| values.lower().err_into())
+                .and_then(|values| {
+                    Optimizer::logical_optimizer(&mz_transform::typecheck::empty_context())
```

Comment:
Not super actionable so feel free to proceed regardless but it does seem strange for the "client" side to care about needing to optimize

Reply:
We either optimize in the coordinator or not in the coordinator, and we want to move as many things as possible out of the coordinator. https://github.com/MaterializeInc/materialize/pull/18994 was recently merged to move peek optimization out. This is another one of those.